pandas/plotting/_matplotlib/hist.py

```diff
@@ -21,22 +21,38 @@ def __init__(self, data, bins=10, bottom=0, **kwargs):
         MPLPlot.__init__(self, data, **kwargs)
 
     def _args_adjust(self):
+
+        # calculate bin number separately in different subplots
+        # where subplots are created based on by argument
         if is_integer(self.bins):
-            # create common bin edge
-            values = self.data._convert(datetime=True)._get_numeric_data()
-            values = np.ravel(values)
-            values = values[~isna(values)]
-
-            _, self.bins = np.histogram(
-                values,
-                bins=self.bins,
-                range=self.kwds.get("range", None),
-                weights=self.kwds.get("weights", None),
-            )
+            if self.by is None:
+                self.bins = self._caculcate_bins(self.data)
+
+            else:
+                grouped = self.data.groupby(self.by)[self.column]
+                bins_list = []
+                for key, group in grouped:
+                    bins_list.append(self._caculcate_bins(group))
+                self.bins = bins_list
 
         if is_list_like(self.bottom):
             self.bottom = np.array(self.bottom)
 
+    def _caculcate_bins(self, data):
```

Comment:
Can you annotate newly created functions?

Reply:
annotated